## NHÓM 11: 
- 19127476 : Trần Thị Huế Minh
- 19127486: Nguyễn Lê Nguyên
- 19127125 : Lê Duy Dũng

# TIỀN XỬ LÝ DỮ LIỆU

In [7]:
import numpy as np
import pandas as pd

## I. Quan sát dữ liệu tổng quát

In [8]:
df = pd.read_csv("covid_data_now.csv")
df

#     Country,Other   TotalCases NewCases  \
0      1               USA  105,547,760   +8,964   
1      2             India   44,689,072      NaN   
2      3            France   39,639,112      NaN   
3      4           Germany   38,241,231      NaN   
4      5            Brazil   37,085,520      NaN   
..   ...               ...          ...      ...   
226  227  Diamond Princess          712      NaN   
227  228      Vatican City           29      NaN   
228  229    Western Sahara           10      NaN   
229  230        MS Zaandam            9      NaN   
230  231           Tokelau            5      NaN   

                                   TotalDeaths  NewDeaths TotalRecovered  \
0    1,148,391                                      129.0    103,083,830   
1      530,776                                        NaN     44,155,119   
2      165,073                                        NaN     39,413,395   
3      168,808                                        NaN     37,823,200   
4      699,310                                        NaN     36,249,161   
..                                         ...        ...            ...   
226         13                                        NaN            699   
227                                                   NaN             29   
228          1                                        NaN              9   
229          2                                        NaN              7   
230                                                   NaN            NaN   

    NewRecovered ActiveCases Serious,Critical Tot Cases/1M pop Deaths/1M pop  \
0        +29,831   1,315,539            2,389          315,251         3,430   
1            NaN       3,177              NaN           31,770           377   
2         +4,021      60,644              869          604,397         2,517   
3        +16,700     249,223              NaN          455,884         2,012   
4            NaN     137,049              NaN          172,208         3,247   
..           ...         ...              ...              ...           ...   
226          NaN           0              NaN              NaN           NaN   
227          NaN           0              NaN           36,295           NaN   
228          NaN           0              NaN               16             2   
229          NaN           0              NaN              NaN           NaN   
230          NaN           5              NaN            3,628           NaN   

        TotalTests Tests/1M pop      Population          Continent  
0    1,168,472,344    3,490,006    334,805,269       North America  
1      919,314,929      653,558  1,406,631,776                Asia  
2      271,490,188    4,139,547     65,584,518              Europe  
3      122,332,384    1,458,359     83,883,596              Europe  
4       63,776,166      296,146    215,353,593       South America  
..             ...          ...             ...                ...  
226            NaN          NaN                                NaN  
227            NaN          NaN            799              Europe  
228            NaN          NaN        626,161              Africa  
229            NaN          NaN                                NaN  
230            NaN          NaN          1,378   Australia/Oceania  

[231 rows x 16 columns]

In [9]:
df.columns

Index(['#', 'Country,Other', 'TotalCases', 'NewCases', 'TotalDeaths',
       'NewDeaths', 'TotalRecovered', 'NewRecovered', 'ActiveCases',
       'Serious,Critical', 'Tot Cases/1M pop', 'Deaths/1M pop', 'TotalTests',
       'Tests/1M pop', 'Population', 'Continent'],
      dtype='object')

### Nhận xét về kiểu dữ liệu:
- Theo quan sát các trường dữ liệu có mô tả thuộc tính như sau:
    - #: STT
    - Country,Other:Tên quốc gia
    - TotalCases: Tổng số lượng ca nhiễm
    - NewCases: Số ca nhiễm mới trong ngày
    - TotalDeaths: Tổng số ca tử vong do Covid-19
    - NewDeaths: Số ca tử vong mới được phát hiện trong ngày
    - TotalRecovered: Tổng số ca hồi phục khi bị nhiễm Covid-19
    - NewRecovered: Số ca hồi phục trong ngày
    - ActiveCases: Số ca đang bị nhiễm và chưa khỏi bệnh
    - Serious,Critical: Số lượng ca nhiễm Covid-19 đang trong nguy kịch
    - Tot Cases/1M pop: Tổng số ca được xác nhận nhiễm trên 1 triệu dân số
    - Deaths/1M pop: Tổng số ca tử vong trên 1 triệu dân số
    - TotalTests: Tổng số lượng được xác nghiệm Covid-19
    - Tests/1M pop: Tổng số lượng được xác nghiệm Covid-19 trên 1 triệu dân số
    - Population: Tổng dân số của quốc gia 
    - Continent: Tên châu lục
   
   #### Trong đó:
       - Các cột 'Country,Other' và 'Continent' có kiểu dữ liệu là chuỗi
       - Các cột còn lại là kiểu số nguyên

### Nhận xét về thành phần quốc gia và châu lục

In [10]:
unique_values_country = df['Country,Other'].unique()
print('Các giá trị duy nhất trong cột Country,Other:')
print(unique_values_country)

print()

num_unique_values_country = df['Country,Other'].nunique()
print('Số lượng giá trị duy nhất trong cột Country,Other:', num_unique_values_country)

Các giá trị duy nhất trong cột Country,Other:
['USA' 'India' 'France' 'Germany' 'Brazil' 'Japan' 'S. Korea' 'Italy' 'UK'
 'Russia' 'Turkey' 'Spain' 'Vietnam' 'Australia' 'Taiwan' 'Argentina'
 'Netherlands' 'Iran' 'Mexico' 'Indonesia' 'Poland' 'Colombia' 'Austria'
 'Greece' 'Portugal' 'Ukraine' 'Chile' 'Malaysia' 'Israel' 'DPRK'
 'Thailand' 'Belgium' 'Czechia' 'Canada' 'Peru' 'Switzerland'
 'Philippines' 'South Africa' 'Romania' 'Denmark' 'Hong Kong' 'Sweden'
 'Serbia' 'Iraq' 'Singapore' 'New Zealand' 'Hungary' 'Bangladesh'
 'Slovakia' 'Georgia' 'Jordan' 'Ireland' 'Pakistan' 'Norway' 'Finland'
 'Kazakhstan' 'Slovenia' 'Lithuania' 'Bulgaria' 'Morocco' 'Croatia'
 'Guatemala' 'Lebanon' 'Costa Rica' 'Bolivia' 'Tunisia' 'Cuba' 'Ecuador'
 'UAE' 'Uruguay' 'Panama' 'Mongolia' 'Nepal' 'Belarus' 'Latvia'
 'Saudi Arabia' 'Azerbaijan' 'Paraguay' 'Bahrain' 'Sri Lanka' 'Kuwait'
 'Dominican Republic' 'Cyprus' 'Myanmar' 'Palestine' 'Estonia' 'Moldova'
 'Venezuela' 'Egypt' 'Libya' 'China' 'Ethiopia' 'Qa

In [11]:
unique_values_continent = df['Continent'].unique()
print('Các giá trị duy nhất trong cột Continent:')
print(unique_values_continent)

print()

num_unique_values_continent = df['Continent'].nunique()
print('Số lượng giá trị duy nhất trong cột Continent:', num_unique_values_continent)

Các giá trị duy nhất trong cột Continent:
['North America' 'Asia' 'Europe' 'South America' 'Australia/Oceania'
 'Africa' nan]

Số lượng giá trị duy nhất trong cột Continent: 6


## II. Tiền xử lý dữ liệu

### 1. Xóa cột không cần thiết
#### Nhận xét: 
   Cột '#' là số thứ tự không cần thiết khi phân tích

In [12]:
df = df.drop('#', axis=1)
df

Country,Other   TotalCases NewCases  \
0                 USA  105,547,760   +8,964   
1               India   44,689,072      NaN   
2              France   39,639,112      NaN   
3             Germany   38,241,231      NaN   
4              Brazil   37,085,520      NaN   
..                ...          ...      ...   
226  Diamond Princess          712      NaN   
227      Vatican City           29      NaN   
228    Western Sahara           10      NaN   
229        MS Zaandam            9      NaN   
230           Tokelau            5      NaN   

                                   TotalDeaths  NewDeaths TotalRecovered  \
0    1,148,391                                      129.0    103,083,830   
1      530,776                                        NaN     44,155,119   
2      165,073                                        NaN     39,413,395   
3      168,808                                        NaN     37,823,200   
4      699,310                                        NaN     36,249,161   
..                                         ...        ...            ...   
226         13                                        NaN            699   
227                                                   NaN             29   
228          1                                        NaN              9   
229          2                                        NaN              7   
230                                                   NaN            NaN   

    NewRecovered ActiveCases Serious,Critical Tot Cases/1M pop Deaths/1M pop  \
0        +29,831   1,315,539            2,389          315,251         3,430   
1            NaN       3,177              NaN           31,770           377   
2         +4,021      60,644              869          604,397         2,517   
3        +16,700     249,223              NaN          455,884         2,012   
4            NaN     137,049              NaN          172,208         3,247   
..           ...         ...              ...              ...           ...   
226          NaN           0              NaN              NaN           NaN   
227          NaN           0              NaN           36,295           NaN   
228          NaN           0              NaN               16             2   
229          NaN           0              NaN              NaN           NaN   
230          NaN           5              NaN            3,628           NaN   

        TotalTests Tests/1M pop      Population          Continent  
0    1,168,472,344    3,490,006    334,805,269       North America  
1      919,314,929      653,558  1,406,631,776                Asia  
2      271,490,188    4,139,547     65,584,518              Europe  
3      122,332,384    1,458,359     83,883,596              Europe  
4       63,776,166      296,146    215,353,593       South America  
..             ...          ...             ...                ...  
226            NaN          NaN                                NaN  
227            NaN          NaN            799              Europe  
228            NaN          NaN        626,161              Africa  
229            NaN          NaN                                NaN  
230            NaN          NaN          1,378   Australia/Oceania  

[231 rows x 15 columns]

In [13]:
# Kiểm tra cột dữ liệu
df.columns

Index(['Country,Other', 'TotalCases', 'NewCases', 'TotalDeaths', 'NewDeaths',
       'TotalRecovered', 'NewRecovered', 'ActiveCases', 'Serious,Critical',
       'Tot Cases/1M pop', 'Deaths/1M pop', 'TotalTests', 'Tests/1M pop',
       'Population', 'Continent'],
      dtype='object')

### 2. Chuyển kiểu dữ liệu

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Country,Other     231 non-null    object 
 1   TotalCases        231 non-null    object 
 2   NewCases          46 non-null     object 
 3   TotalDeaths       231 non-null    object 
 4   NewDeaths         23 non-null     float64
 5   TotalRecovered    210 non-null    object 
 6   NewRecovered      50 non-null     object 
 7   ActiveCases       212 non-null    object 
 8   Serious,Critical  127 non-null    object 
 9   Tot Cases/1M pop  229 non-null    object 
 10  Deaths/1M pop     223 non-null    object 
 11  TotalTests        213 non-null    object 
 12  Tests/1M pop      213 non-null    object 
 13  Population        231 non-null    object 
 14  Continent         229 non-null    object 
dtypes: float64(1), object(14)
memory usage: 27.2+ KB


#### Nhận xét : 
- Trong mô tả dữ liệu trên:
    - Chỉ có cột `NewDeaths` là có kiểu dữ liệu `float64`
    - Các cột còn lại có kiểu là `object`
- Để phân tích và thống dữ liệu ta cần chuyển kiểu dữ liệu của các cột như đã nhận xét trong phần Tổng quan như sau:
    - Cột `Country,Other` và `Continent` có kiểu dữ liệu là `string`
    - Các cột còn lại có kiểu là `integer`

In [15]:
df["Country,Other"] = df["Country,Other"].astype('string')
df["Continent"] = df["Continent"].astype('string')

In [16]:
col = ["Country,Other", "Continent"]

for c in df.columns:
    if c in col:
        continue
    else:
        df[c] = df[c].astype(str)
        df[c] = df[c].str.replace(',', '')
        df[c] = pd.to_numeric(df[c], errors='coerce').astype('Int64') 

In [19]:
# Kiểm tra lại kiểu dữ liệu
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Country,Other     231 non-null    string
 1   TotalCases        231 non-null    Int64 
 2   NewCases          46 non-null     Int64 
 3   TotalDeaths       225 non-null    Int64 
 4   NewDeaths         23 non-null     Int64 
 5   TotalRecovered    210 non-null    Int64 
 6   NewRecovered      50 non-null     Int64 
 7   ActiveCases       212 non-null    Int64 
 8   Serious,Critical  127 non-null    Int64 
 9   Tot Cases/1M pop  229 non-null    Int64 
 10  Deaths/1M pop     223 non-null    Int64 
 11  TotalTests        213 non-null    Int64 
 12  Tests/1M pop      213 non-null    Int64 
 13  Population        229 non-null    Int64 
 14  Continent         229 non-null    string
dtypes: Int64(13), string(2)
memory usage: 30.1 KB


### 3. Kiểm tra và xử lý missing value

#### Số lượng missing value trong mỗi cột thuộc tính và tỉ lệ phần trăm của giá trị missing

In [20]:
count_missing = df.isnull().sum()
percent_missing = round(df.isnull().sum() * 100 / len(df),3)
df_missing = count_missing.to_frame(name='Mumbers')
df_missing.insert(loc=1, column='Percent missing', value=percent_missing)
df_missing

Mumbers  Percent missing
Country,Other           0            0.000
TotalCases              0            0.000
NewCases              185           80.087
TotalDeaths             6            2.597
NewDeaths             208           90.043
TotalRecovered         21            9.091
NewRecovered          181           78.355
ActiveCases            19            8.225
Serious,Critical      104           45.022
Tot Cases/1M pop        2            0.866
Deaths/1M pop           8            3.463
TotalTests             18            7.792
Tests/1M pop           18            7.792
Population              2            0.866
Continent               2            0.866

##### Nhận xét
   - Phần lớn các cột đều có giá trị null. Trong đó:
   
       - Các cột có tỉ lệ missing value cao hơn 70% là:
            - `NewCases` :           80,087%
            - `NewDeaths` :          90,043%
            - `NewRecovered` :      78.355%
            
       - Các cột có tỉ lệ missing value trung bình là:
            - `Serious,Critical` :  45.022%
            
       - Các cột có tỉ lệ missing value thấp dưới 10% là:
            - `TotalRecovered` :     9.091%
            - `ActiveCases` :        8.225%
            - `Deaths/1M pop` :      3.463%
            - `TotalTests` :         7.792%
            - `Tests/1M pop` :       7.792%
            - `Continent` :          0.866%
            - `Tot Cases/1M pop` :   0.866%
            
       - Các cột không có missing value là: 
           - ````Country,Other````, ````TotalCases````, ````TotalDeaths````, ````Population````	

#### Xử lý missing value
 * Đối với các cột dữ liệu có tỉ lệ missing value cao hơn 70%
     - Dữ liệu missing ở các cột này khá cao, đồng thời những dữ liệu này cho biết số ca nhiễm, số ca tử vong, số ca hồi phục mới trong ngày và những số liệu này không đóng góp nhiều vào quá trình phân tích và thống kê --> Chọn loại bỏ các cột dữ liệu này
        

In [21]:
df = df.drop(columns=["NewCases", "NewRecovered", "NewDeaths"])
df

Country,Other  TotalCases  TotalDeaths  TotalRecovered  ActiveCases  \
0                 USA   105547760      1148391       103083830      1315539   
1               India    44689072       530776        44155119         3177   
2              France    39639112       165073        39413395        60644   
3             Germany    38241231       168808        37823200       249223   
4              Brazil    37085520       699310        36249161       137049   
..                ...         ...          ...             ...          ...   
226  Diamond Princess         712           13             699            0   
227      Vatican City          29         <NA>              29            0   
228    Western Sahara          10            1               9            0   
229        MS Zaandam           9            2               7            0   
230           Tokelau           5         <NA>            <NA>            5   

     Serious,Critical  Tot Cases/1M pop  Deaths/1M pop  TotalTests  \
0                2389            315251           3430  1168472344   
1                <NA>             31770            377   919314929   
2                 869            604397           2517   271490188   
3                <NA>            455884           2012   122332384   
4                <NA>            172208           3247    63776166   
..                ...               ...            ...         ...   
226              <NA>              <NA>           <NA>        <NA>   
227              <NA>             36295           <NA>        <NA>   
228              <NA>                16              2        <NA>   
229              <NA>              <NA>           <NA>        <NA>   
230              <NA>              3628           <NA>        <NA>   

     Tests/1M pop  Population          Continent  
0         3490006   334805269      North America  
1          653558  1406631776               Asia  
2         4139547    65584518             Europe  
3         1458359    83883596             Europe  
4          296146   215353593      South America  
..            ...         ...                ...  
226          <NA>        <NA>               <NA>  
227          <NA>         799             Europe  
228          <NA>      626161             Africa  
229          <NA>        <NA>               <NA>  
230          <NA>        1378  Australia/Oceania  

[231 rows x 12 columns]

#### Xử lý missing value
 * Đối với các cột dữ liệu có tỉ lệ missing value trung bình

In [22]:
country_Serious = []
for i in range(0, len(df["Serious,Critical"])):
    if df.isnull().iloc[i, 5]:
        country_Serious.append(df.loc[i, "Country,Other"])
print(country_Serious)

['India', 'Germany', 'Brazil', 'UK', 'Russia', 'Turkey', 'Taiwan', 'Mexico', 'Indonesia', 'Ukraine', 'DPRK', 'Thailand', 'New Zealand', 'Bangladesh', 'Georgia', 'UAE', 'Nepal', 'Belarus', 'Azerbaijan', 'Sri Lanka', 'Kuwait', 'Myanmar', 'Libya', 'China', 'Armenia', 'Bosnia and Herzegovina', 'North Macedonia', 'Zambia', 'Albania', 'Luxembourg', 'Algeria', 'Laos', 'Namibia', 'Jamaica', 'Cambodia', 'Rwanda', 'Barbados', 'Angola', 'Channel Islands', 'DRC', 'Senegal', 'Ivory Coast', 'Suriname', 'Eswatini', 'Guyana', 'Fiji', 'Madagascar', 'Sudan', 'Mauritania', 'Bhutan', 'Syria', 'Burundi', 'Seychelles', 'Gabon', 'Mayotte', 'Mauritius', 'Togo', 'Isle of Man', 'Lesotho', 'Haiti', 'Mali', 'Saint Lucia', 'Benin', 'Somalia', 'Congo', 'Timor-Leste', 'Burkina Faso', 'Gibraltar', 'Bermuda', 'Nicaragua', 'Tajikistan', 'Tonga', 'Samoa', 'Dominica', 'Djibouti', 'Marshall Islands', 'Gambia', 'Vanuatu', 'Caribbean Netherlands', 'Sint Maarten', 'Eritrea', 'St. Vincent Grenadines', 'Comoros', 'Sierra Leone

- Nhận xét:
    - Vì số lượng missing value trong cột Serious,Critical cũng ở mức cao, nhưng cột dữ liệu này có ý nghĩa thống kê quan trọng vì vậy nhóm chọn phương pháp điền giá trị trống trong cột này là 0 để mục đích khi phân tích có thể so sánh được mức độ Serious,Critical của các quốc gia 

In [23]:
df["Serious,Critical"] = df["Serious,Critical"].fillna(0)
df

Country,Other  TotalCases  TotalDeaths  TotalRecovered  ActiveCases  \
0                 USA   105547760      1148391       103083830      1315539   
1               India    44689072       530776        44155119         3177   
2              France    39639112       165073        39413395        60644   
3             Germany    38241231       168808        37823200       249223   
4              Brazil    37085520       699310        36249161       137049   
..                ...         ...          ...             ...          ...   
226  Diamond Princess         712           13             699            0   
227      Vatican City          29         <NA>              29            0   
228    Western Sahara          10            1               9            0   
229        MS Zaandam           9            2               7            0   
230           Tokelau           5         <NA>            <NA>            5   

     Serious,Critical  Tot Cases/1M pop  Deaths/1M pop  TotalTests  \
0                2389            315251           3430  1168472344   
1                   0             31770            377   919314929   
2                 869            604397           2517   271490188   
3                   0            455884           2012   122332384   
4                   0            172208           3247    63776166   
..                ...               ...            ...         ...   
226                 0              <NA>           <NA>        <NA>   
227                 0             36295           <NA>        <NA>   
228                 0                16              2        <NA>   
229                 0              <NA>           <NA>        <NA>   
230                 0              3628           <NA>        <NA>   

     Tests/1M pop  Population          Continent  
0         3490006   334805269      North America  
1          653558  1406631776               Asia  
2         4139547    65584518             Europe  
3         1458359    83883596             Europe  
4          296146   215353593      South America  
..            ...         ...                ...  
226          <NA>        <NA>               <NA>  
227          <NA>         799             Europe  
228          <NA>      626161             Africa  
229          <NA>        <NA>               <NA>  
230          <NA>        1378  Australia/Oceania  

[231 rows x 12 columns]

#### Xử lý missing value
 * Đối với các cột dữ liệu có tỉ lệ missing value thấp:
     - Nhận thấy đối với 2 cặp cột dữ liệu [`Continent`,`Tot Cases/1M pop`] và [`TotalTests` ; `Tests/1M pop`] có tỉ lệ missing value bằng nhau lần lượt là 0.866% và 7.792%
     - Đặt giả thuyết rằng các giá trị missing tại các cặp cột này là của cùng một quốc gia. Ta kiểm tra như sau: 

In [24]:
country_Continent = []

for i in df.index:

    try:
        if df.isnull().iloc[i, 11]:
            country_Continent.append(i)
            print('Continent _', i, ': ', df.loc[i, "Country,Other"])
    except:
        None
    
    try:
        if df.isnull().iloc[i, 6]:
            print('Tot Cases/1M pop _', i, ': ', df.loc[i, "Country,Other"])
    except:
        None

Continent _ 226 :  Diamond Princess
Tot Cases/1M pop _ 226 :  Diamond Princess
Continent _ 229 :  MS Zaandam
Tot Cases/1M pop _ 229 :  MS Zaandam


##### Nhận xét:
- Đúng với giả thuyết các ô dữ liệu bị thiếu thuộc cùng một Quốc gia
    - Tại bộ [`Continent`,`Tot Cases/1M pop`] cột dữ liệu `Continent` xác định Châu lục mà quốc gia đó thuộc về bị null, điều này có thể gây ảnh hưởng đến việc phân tích và thống kê nếu điền giá trị null bằng phương pháp thay thế. Vì vậy nhóm lựa chọn phương án loại bỏ các sample này và xem chúng như outlier của bộ dữ liệu

In [25]:
df = df.drop(country_Continent, axis=0).reset_index(drop=True)
print("Số sample và thuộc tính còn lại:", df.shape)

Số sample và thuộc tính còn lại: (229, 12)


In [26]:
country_TotalTests = []
country_Tests1Mpop = []
for i in df.index:

    try:
        if df.isnull().iloc[i, 8]:
            country_TotalTests.append(i)
            print('TotalTests _', i, ': ', df.loc[i, "Country,Other"], ': ',  df.loc[i, "Population"])
    except:
        None
    
    try:
        if df.isnull().iloc[i, 9]:
            print('Tests1Mpop _', i, ': ', df.loc[i, "Country,Other"], ': ',  df.loc[i, "Population"])
    except:
        None

TotalTests _ 29 :  DPRK :  25990679
Tests1Mpop _ 29 :  DPRK :  25990679
TotalTests _ 138 :  French Polynesia :  284164
Tests1Mpop _ 138 :  French Polynesia :  284164
TotalTests _ 150 :  Seychelles :  99426
Tests1Mpop _ 150 :  Seychelles :  99426
TotalTests _ 156 :  Tanzania :  63298550
Tests1Mpop _ 156 :  Tanzania :  63298550
TotalTests _ 172 :  Solomon Islands :  721159
Tests1Mpop _ 172 :  Solomon Islands :  721159
TotalTests _ 181 :  Nicaragua :  6779100
Tests1Mpop _ 181 :  Nicaragua :  6779100
TotalTests _ 183 :  Tajikistan :  9957464
Tests1Mpop _ 183 :  Tajikistan :  9957464
TotalTests _ 190 :  Marshall Islands :  60057
Tests1Mpop _ 190 :  Marshall Islands :  60057
TotalTests _ 203 :  Comoros :  907419
Tests1Mpop _ 203 :  Comoros :  907419
TotalTests _ 216 :  Kiribati :  123419
Tests1Mpop _ 216 :  Kiribati :  123419
TotalTests _ 221 :  Tuvalu :  12066
Tests1Mpop _ 221 :  Tuvalu :  12066
TotalTests _ 222 :  Saint Helena :  6115
Tests1Mpop _ 222 :  Saint Helena :  6115
TotalTests _ 2

##### Nhận xét:
- Đúng với giả thuyết các ô dữ liệu bị thiếu thuộc cùng một Quốc gia
    - Tại bộ [`TotalTests` ; `Tests/1M pop`] , Các sample bị thiếu thuộc các quốc gia có số lượng dân số thấp, và nhận thấy rằng nếu loại bỏ các sample này sẽ không làm ảnh hưởng quá nhiều đến việc thống kê sau này, bên cạnh đó còn hạn chế được các giá trị missing đáng kể

In [27]:
df = df.drop(country_TotalTests, axis=0).reset_index(drop=True)
print("Số sample và thuộc tính còn lại:", df.shape)

Số sample và thuộc tính còn lại: (213, 12)


#### Kiểm tra lại phần trăm missing value

In [28]:
count_missing = df.isnull().sum()
percent_missing = round(df.isnull().sum() * 100 / len(df),3)
df_missing = count_missing.to_frame(name='Mumbers')
df_missing.insert(loc=1, column='Percent missing', value=percent_missing)
df_missing

Mumbers  Percent missing
Country,Other           0            0.000
TotalCases              0            0.000
TotalDeaths             1            0.469
TotalRecovered         16            7.512
ActiveCases            16            7.512
Serious,Critical        0            0.000
Tot Cases/1M pop        0            0.000
Deaths/1M pop           1            0.469
TotalTests              0            0.000
Tests/1M pop            0            0.000
Population              0            0.000
Continent               0            0.000

##### Nhận xét:
- Các cột còn lại có missing value khá thấp, nhóm lựa chọn phương pháp giữ nguyên các dữ liệu đó để tránh làm sai lệch dữ liệu quá nhiều

## III. Kiểm tra tổng quát

In [29]:
df

Country,Other  TotalCases  TotalDeaths  TotalRecovered  \
0                      USA   105547760      1148391       103083830   
1                    India    44689072       530776        44155119   
2                   France    39639112       165073        39413395   
3                  Germany    38241231       168808        37823200   
4                   Brazil    37085520       699310        36249161   
..                     ...         ...          ...             ...   
208                  Macao        3514          121            3392   
209  Saint Pierre Miquelon        3452            2            2449   
210      Wallis and Futuna        3427            7             438   
211       Falkland Islands        1930         <NA>            1930   
212             Montserrat        1403            8            1376   

     ActiveCases  Serious,Critical  Tot Cases/1M pop  Deaths/1M pop  \
0        1315539              2389            315251           3430   
1           3177                 0             31770            377   
2          60644               869            604397           2517   
3         249223                 0            455884           2012   
4         137049                 0            172208           3247   
..           ...               ...               ...            ...   
208            1                 0              5264            181   
209         1001                 1            599410            347   
210         2982                 0            312056            637   
211            0                 0            545352           <NA>   
212           19                 0            282578           1611   

     TotalTests  Tests/1M pop  Population          Continent  
0    1168472344       3490006   334805269      North America  
1     919314929        653558  1406631776               Asia  
2     271490188       4139547    65584518             Europe  
3     122332384       1458359    83883596             Europe  
4      63776166        296146   215353593      South America  
..          ...           ...         ...                ...  
208        7850         11760      667490               Asia  
209       25400       4410488        5759      North America  
210       20508       1867419       10982  Australia/Oceania  
211        8632       2439107        3539      South America  
212       17762       3577442        4965      North America  

[213 rows x 12 columns]

In [30]:
print("Số lượng dòng:", df.shape[0])
print("Số lượng cột:", df.shape[1])
print("Số lượng quốc gia:", df['Country,Other'].nunique())
print("Số lượng châu lục:", df['Continent'].nunique())

Số lượng dòng: 213
Số lượng cột: 12
Số lượng quốc gia: 213
Số lượng châu lục: 6


## VI. Xuất dữ liệu mới

In [234]:
df.to_csv('new_covid_data_now.csv', index=False)